### PR TITLE
boards: move shared code to boards/common/xx

### PR DESCRIPTION
**Pull request by haukepetersen**
As discussed/proposed in #8044: this PR puts all shared board code under `/boards/common/xx/` instead of using shared folders as in `boards/xx-common/`.

Quote:
> This allows us to use shard code among boards, that are not logically connected to each other (e.g. not part of one family or revisions of each other etc). So for example we could add boards/common/stm and put STM specific clock configurations (like periph_clk_f4_180.h in this PR) into this 'module'. These configuration snippets can then simply be used from any board!
>
> Right now this would mean to put a stm-common folder into boards, which might look a little strange, and I think having a structured boards/common/xx place for configuration snippets might be cleaner.

### CODE REVIEW

**Comment by cladmi** [comment]
Do we want to make the common boards in a separate group for the documentation, like `board_common` ? This way it could be displayed separately in the documentation http://doc.riot-os.org/group__boards.html

If it is wanted, I can try something after this PR is merged.

**Comment by haukepetersen** [comment]
> If it is wanted, I can try something after this PR is merged.

Yes, I think this would be a great idea. I intentionally did not touch any doxygen grouping with this PR, so adding that on top would be my preferred approach.

**Comment by cladmi** [comment]
Is it required to write commits so that every commits work? Or at least that we try to.

I ask because right now, `nucleo144` and `nucleo32` commits will not work as `nucleo-common` is migrated to `common/nucleo` in a following commit.

And also, things will be broken before the last fix commits.

**Comment by kaspar030** [comment]
> Is it required to write commits so that every commits work? Or at least that we try to.

We have the (unwritten) rule that commits must work on PR merge commit level. So, if some commits within one PR don't build unless some other commits of the same PR get merged, that's fine.

File: boards/common/iotlab/Makefile [hunk 1]
```diff
@@ -1,3 +1,3 @@
-MODULE = arduino-mkr-common
```
Comment (kaspar030) [inline boards/common/iotlab/Makefile]: next time use "git mv" ;)

**Comment by kaspar030** [comment]
pre-ACK. We should give a change of this magnitude a little more time. @gebart @miri64 @smlng what do you think?

**Comment by haukepetersen** [comment]
> We should give a change of this magnitude a little more time

+1

**Comment by aabadie** [comment]
Thinking of it, I would also add a `boards/common/stm`: there we can put what's common to all stm based board. This directory could be the base not only for nucleo but also the stm32discovery, the b-l072z-lrwan1 (I hate this name), bluepill, the one that will come with #7585, etc
The same idea could apply to sam21 based board (arduino-zero, sodaq-*, arduino-mkr*)

What do you think ?

**Comment by miri64** [comment]
> While the (semantically) better approach (I suggest) would be, to group similar boards into one subdirectory which would have a common folder, i.e., the above would like:

I rhat don't like that proposal. I think it is really nice that you basically can get all supported boards by doing

```sh
ls boards/ | grep -v "common"
```

and the current proposal makes that even easier, since you can basically ditch the `grep -v "common"` and just ignore the `common` directory "manually".

With your proposal that would be more complicated:

```sh
ls boards/*/ | grep -v "common"
```

I don't think it's more or less logical logical than the one @haukepetersen proposed. Your suggestion implies that all stuff is at its next higher hierarchy (both common and specific), @haukepetersen's approach implies that all common things are in one place, ordered by the next higher hierarchy.

**Comment by haukepetersen** [comment]
@smlng I don't agree for multiple reasons:

1. the list of supported board (as in one folder per board) should be in a single, flat directory. Everything else would add unnecessary complexity to the build system, but would also IMHO make a developer's life (especially for non-core-people) harder, as they would have no clue where there board implementation resides -> simplicity in this case is king! With a flat directory containting all the boards, you would always have a starting point from where one can deduct all the rest of shared modules. Also keeping the flat top-level board structure is non-intrusive and does not require anything changed in the build system.

2. (and this is my much stronger argument) **Boards and their shared code don't have a strict hierarchical structure!** As this is mostly the case for our supported CPUs (`CPU_ARCH -> CPU_FAM -> CPU -> CPU_MODEL`), this is not the case for common board code. Sure we could map something like `stm32 -> nucleo -> nucleo144 -> nucleo144-f746` or `sam0 -> arduino-mkr -> ardunino-mkr1000` to a sub-folder structure, but what if we also want to introduce a sub-module, that is used by all `stm32` and `sam0` CPUs? This would not fit in this structure!

So looking at
```
boards/stm32/common
boards/stm32/nucleo/common/nucleo144/
boards/stm32/nucleo/nucleo144-f746/
boards/stm32/opencm904/

boards/sam0/common
boards/sam0/arduino-mkr/common
boards/sam0/arduino-mkr/arduino-mkr1000
boards/sam0/arduino-mkr/arduino-mkrzero
```
compared to
```
boards/nucleo144-f746
boards/opencm904
boards/arduino-mkr1000
boards/arduino-mkrzero
boards/common/stm
boards/common/nucleo
boards/common/nucleo144
boards/common/sam0
boards/common/arduino-mkr
```
I'd am strongly for the second, flat, and simple structure!

On top, keeping the structure flat and simple, it means that we can easily create 'vertical' shared modules, that don't fit in the context of a certain board or cpu family. So for example we could easily add a module like `boards/common/tools` and `boards/common/tools` and put in specific Makfiles for the 2 prevalent serial configuration that we encounter everywhere in the board code.

**Comment by gebart** [comment]
I like this proposal. Especially being able to create common clock configurations for all boards sharing a single CPU will reduce code duplication a lot.
I agree with @miri64 and @haukepetersen regarding the structure, it makes it easier to get an overview rather than to have to search for names

**Comment by smlng** [comment]
> Or in less words -> 'What @miri64 said` :-)

Not really -- @miri64 mainly _complained_ that my suggested approach would make it harder to use `grep`, while you put up some more valid arguments. However, on the one hand you over exaggerated my approach and on the other you also started to mix CPU/MCU and boards, i.e., there is no `sam0_common` folder in boards right now -- or do you plan to introduce one? At least I wasn't. 

For the former: my idea wasn't to introduce an arbitrarily deep directory structure for boards, but rather group those boards that already have a `<board-prefix>_common` folder into `<board-prefix>/` with subdirs `<board-prefix>/common`, `<board-prefix>/<board1>`, ..., `<board-prefix>/<boardN>`. Mapping this onto the current board structure is as easy as your approach, besides that the nucleos are poorly/inconsistently named -> `nucleo-*` should be `nucleo64-*` to match with `nucleo32-*` and `nucleo144-*`.

Anyhow, I agree with you that simplicity is key, nevertheless in the long run a little structure might also help. My points is, that we should not hastily restructure boards without considering and discussing alternatives - hence my suggestion, to show there are other possibilities.

**Comment by miri64** [comment]
> Not really -- @miri64 mainly complained that my suggested approach would make it harder to use grep

I also spoke about the hierarchical implications. And I said that with this approach I won't even need grep ;-).

**Comment by smlng** [comment]
> it makes it easier to get an overview rather than to have to search for names

so you think searching in a flat hierarchy is faster than traversing a (shallow) tree 😉 

Assuming you have a random `nucleo144-*` or want to buy one and see which one is supported by RIOT. With @haukepetersen approach @miri64 would have to `ls boards/ | grep nucleo144` and with my suggestion this becomes `ls boards/nucleo144/`. 

Though I admit, getting an overview over all boards is a bit more complicated  as @miri64 showed above - you have to use `boards/*` to find all 😮

**Comment by gebart** [comment]
@smlng 
>there is no sam0_common folder in boards right now -- or do you plan to introduce one? At least I wasn't.

I intend to introduce a kinetis directory for the kinetis clock configuration which will be shared between all Kinetis boards. My previous proposal was to put it in cpu, but you suggested that it should not be in cpu but under boards, so with this method in combination with the same ideas in #8044 we can reduce code duplication.

**Comment by smlng** [comment]
@gebart I understand, maybe we should also consider #8066 here?

**Comment by miri64** [comment]
> With @haukepetersen approach @miri64 would have to ls boards/ | grep nucleo144 and with my suggestion this becomes ls boards/nucleo144/.

Nope, far to complicated :P I would just use `ls boards/nucleo144*`, and I usually try to get an overview over *all* boards than trying to single out a specific board (with boards like the PhyNode I would otherwise be lost ;-)).

**Comment by smlng** [comment]
> with boards like the PhyNode I would otherwise be lost ;-))

well that one is specifically badly named - I think we all agree and @haukepetersen already proposed a solution aka board-aliases #7976

**Comment by haukepetersen** [comment]
> My points is, that we should not hastily restructure boards without considering and discussing alternatives - hence my suggestion, to show there are other possibilities.

That is the point of this PR :-) -> Agreed upon quote from @kaspar030 above `We should give a change of this magnitude a little more time`...

> However, on the one hand you over exaggerated my approach and on the other you also started to mix CPU/MCU and boards, i.e., there is no sam0_common folder in boards right now -- or do you plan to introduce one? At least I wasn't.

Exaggerating was my intention here. My point is, that simply grouping some obvious groups is not enough, also introducing some more fine grained shared modules will be needed/helpful (as pointed out by @gebart). So having some shareable, CPU-bound board configurations (e.g. clock config) will be very handy. But to this point, I'd say we don't know yet which other type of groups will be useful (or will become useful some point in the future), so the 'grouping system' should be able to also cope with 'vertical' groups.

**Comment by smlng** [comment]
> So having some shareable, CPU-bound board configurations (e.g. clock config) will be very handy.

As pointed out in #8066 and others raised that concern, too (e.g. @A-Paul) that we should carefully distinguish board and CPU stuff and where to put what. I'm not sure on that matter either ...

Regarding the clock configuration, I argued to @gebart that such should reside in periph_configs  with board(s) and not with the CPU. Thinking about it again, clock configuration depends on the specific CPU variant placed on a certain board, but it does not depend on the board itself, right? See for instance the `limifrog-v1` and `nz32-sc151` they share the same cpu `stm32l151rc` with the same clock configuration, which would be their `common` board folder to deduplicate the clock config? I think there is no need to have a common board folder for that, but rather move the clock config into the CPU folder.
 
IMHO getting board and CPU configs sorted out and structured, will render the need for your so called `vertical` groups obsolete.

**Comment by haukepetersen** [comment]
>  Thinking about it again, clock configuration depends on the specific CPU variant placed on a certain board, but it does not depend on the board itself, right? 

In general: wrong. The clock configuration depends on things like external oscillators, which are clearly dependent on the board. But everything that is not dependent on the boards should of course not be part of the board config (e.g. the clock speed (`CLOCK_CORECLOCK`) for the `nrf51` CPUs is fixed, so it is defined in `cpu/nrf52/include/periph_cpu.h`).

> IMHO getting board and CPU configs sorted out and structured,

As just stated in #8066, I was under the impression that we do so.

**Comment by gebart** [comment]
@smlng 
>Thinking about it again, clock configuration depends on the specific CPU variant placed on a certain board, but it does not depend on the board itself, right? See for instance the limifrog-v1 and nz32-sc151 they share the same cpu stm32l151rc with the same clock configuration, which would be their common board folder to deduplicate the clock config? I think there is no need to have a common board folder for that, but rather move the clock config into the CPU folder.

Sounds like you changed your mind. I also think it makes sense to put it in the CPU dir, but write the definitions/implementations in a way that it is sensitive to overrides from the board config.

Like @haukepetersen wrote, the only thing that will be board dependent for clock configuration is the presence/absence of external crystals or clock sources, and the nominal frequency of these components, to be able to specify exact target frequencies like UART baud rate or timer frequency.
My vision is to be able to specify any crystals or clock sources present on the board and the CPU clocking mode and target frequency, and get the clock configuration to work out all the dividers and multipliers by itself. Anyway, this could be getting off track for this thread.

**Comment by kYc0o** [comment]
IMHO something is quite clear: anything soldered to the board (crystals, buttons, radios, LEDs, etc...) should belong to board configuration. In the specific case of the clocks, I agree with @gebart that these should be configurable at the cpu level (on which we found PLL, or other new technologies), thus, a "pre-configuration" at the board level should be followed by a definite configuration at the CPU level.

My 2c.

**Comment by haukepetersen** [comment]
>  the only thing that will be board dependent for clock configuration is the presence/absence of external crystals or clock sources, and the nominal frequency of these components

Almost: there is also the case that one wants to run with a reduced clock (as clock speed has a great impact on energy consumption). So say company xyz adds support for their board to RIOT, but they want to run per default at 16MHz coreclock instead of the maximum possible. So this needs to be configurable by the board.

> My vision is to be able to specify any crystals or clock sources present on the board and the CPU clocking mode and target frequency, and get the clock configuration to work out all the dividers and multipliers by itself. Anyway, this could be getting off track for this thread.

+1, this would be the ideal case. We already took some effort in simplifying the CPU configuration needed for certain CPUs. Though the result is always a trade-off between user-friendly config interface, run-time overhead, external tools needed, etc... So IMHO for certain scenarios is is justified to have a couple of additional config options to set, if this does cut down clock init runtime (as we call this code every time we wake up from deep sleep...). But this has to be handled on a case-by-case basis, and here is indeed room for improvement!

Making at least use of shared clock configurations between boards for now would already improve the situation quite a bit...

**Comment by haukepetersen** [comment]
As we drifting away here: this PR is about the place to put common board elements, not how to define the clock configuration... So could we focus on the former (again)?

**Comment by temmihoo** [comment]
### Short note on naming the Nucleo boards:

The Nucleo family has three subfamilies, nucleo32, nucleo64 and nucleo144, two of which are currently visible as named entities and the nucleo64 is known just as nucleo. I think these should be renamed consistently so nucleo-f401 becomes nucleo64-f401 and so on.

### More disputable opinion text on hierarchy below:

Along with @smlng, quoting @haukepetersen, of the hierarchial tree structure I personally would favour the first
  boards/stm32/common
  boards/stm32/nucleo/common/nucleo144/
  boards/stm32/nucleo/nucleo144-f746/
  boards/stm32/opencm904/
  boards/sam0/common
  boards/sam0/arduino-mkr/common
  boards/sam0/arduino-mkr/arduino-mkr1000
  boards/sam0/arduino-mkr/arduino-mkrzero

as opposed to
  boards/nucleo144-f746
  boards/opencm904
  boards/arduino-mkr1000
  boards/arduino-mkrzero
  boards/common/stm
  boards/common/nucleo
  boards/common/nucleo144
  boards/common/sam0
  boards/common/arduino-mkr

Along with this preference I think we should have in the filesystem visible to ls and grep aliases so that the devices can be found using the board vendor, MCU vendor, board family and possibly some other aliases as well. What would be the _real_ name would in my opinion be that by board vendor, making boards/arduino/arduino-mkr/arduino-mkrzero the target of symlink boards/atmel/sam0/arduino-mkr/arduino-mkrzero and so on.

There is more to it. In the above copypasted part the naming scheme is somewhat different to what I discuss since the grouping is by MCU family instead of board vendor. For saving the space I will not examplify this and consider this opinion something to hopefully be discussed further before going into details.

I would not put a strict semantic definition on what a specific hierarchial level means since both board and MCU vendors have their own semantics and we either somehow reflect any of those or none... or risk create the 15th standard[1]. We do not need to make the .mk include script to _understand_ the tree structure if we just write it down in the include directive literally as path relative to RIOTBASE or RIOTBOARD (this is happening already so no change to existing pattern). I have not tested but I am not seeing anything go boingk if we create symlinks to board files and provide users multiple views of the same tree data or that there are varying levels of tree structure for different boards. Only thing that will break is the pattern of referring to ../something which is anyway a bad and confusing pattern.

  [1]: https://xkcd.com/927/

**Comment by kaspar030** [comment]
I think we should tackle this by requirements. I don't think that having the "most logical order" is a requirement.

What I do see as requirements are:

- finding a board's folder is dead simple
- code / configuration can easily be shared between similar boards
- newcomers understand the ordering as quickly as possible
- whatever system we have is "easy to handle"

We currently have that, mostly. The other proposals using a more deeply nested folder structure will make it harder to find a specific board, and make it harder to do the equivalent of "ls boards | grep -v common", and thus will make it easier for newcomers.

This PR just moves the common folders into common/, so the growing number of common folders don't clutter the board directory. I strongly suggest going with this PR, as it essentially keeps the flat hirarchy and only improves what might become messy.

**Comment by kaspar030** [comment]
> This PR just moves the common folders into common/, so the growing number of common folders don't clutter the board directory. I strongly suggest going with this PR, as it essentially keeps the flat hirarchy and only improves what might become messy.

@smlng @kYc0o @temmihoo @gebart What do you think? Take this (maybe slight) improvement for now and postpone further refinement of the (common) folder structure?
There are a couple of PR's that would benefit from this (#8035 #7738 #8047).

**Comment by haukepetersen** [comment]
I have spend some more thought about this discussion and I'd like to make a proposal. As it seems to me, one important argument (aside from complexity) against a more sophisticated re-structuring as proposed by @smlng and @temmihoo is, that we can't fully oversee the actual sub-module structure that we will move to, as we have not started to heavily make use of re-usable board components. Now this is IMHO something best achieved by doing-and-learning.

Now the changes proposed in this PR and non-intrusive (at least from the build systems point of view), and already improve the current situation quite a bit. So I would propose to go with this simple structure for now, do a lot of unification, cleanup, and condensation on top of this, and then come back to the discussion about a more sophisticated structure once we have a very good feeling about the structure about re-used components on this level. Moving from this simple structure to something with deeper structure should be dead simple at any point in time...

**Comment by kYc0o** [comment]
IMHO this is a "small" though important step forward the reorganisation of the tree. The actual status is by far worse than what this PR proposes, thus I tend to agree that getting this in should improve the situation.

I have no big concerns but having a clean and easy way to add new platforms, since I heard critics to other OS on which adding something new is a pain. I don't want this for RIOT, and I think this PR doesn't make things worse, but better.

Though, it needs rebase.

**Comment by haukepetersen** [comment]
will rebased once agreement is reached.

**Comment by smlng** [comment]
As already stated above, I agree that simplicity is king and even though my suggested structure was different from this one here. My idea wasn't to have a deeply nested structure as suggested by @temmihoo (which also seems to mix boards and MCU/CPU), but rather to have *one* additional subdirectory level to groups boards that already have `<something>_common`. However I also had time to think about it, and a drawback of my approach would be that boards  like `z1` would stay in the root directory `boards/z1` while others like the nucleo would be in  a subdirectory.

TL;DR: @haukepetersen go ahead, as this surely improves the situation and might help to clear the view for a _better_ (if there is such) solution/structure in the future.

**Comment by haukepetersen** [comment]
@smlng thanks for your reply, and let's just see what insights some iterations of de-duplication bring us!

**Comment by haukepetersen** [comment]
rebased

**Comment by cladmi** [comment]
When checking the modifications, I was looking for changes that concern `boards/common/nucleo/Makefile` but in fact its never executed, and never was before moving. But that is another topic.

**Comment by temmihoo** [comment]
I agree with @haukepetersen that to make these kinds of restructurings we will learn by doing and in THAT, this pull request is a good step.

I am not certain how deep directory tree would be reasonable and meaningful so I was purposefully being explicit in following that idea to its logical end. That kind of pedantry also necessitates provising aliases for the boards as many of them are known by different names.

Certainly providing aliases by grouping boards by not only by board vendor but MCU family would be good for many purposes. I am not seeing a technical problem with tooling as to my knowledge nothing in the RIOT toolchain traverses the boards/* tree by the file tree. If there are such things, just make them ignore all symlinks.

**Comment by gebart** [comment]
I agree, let's do this PR first, then look at further optimizations.

**Comment by haukepetersen** [comment]
> That kind of pedantry also necessitates provising aliases for the boards as many of them are known by different names.

Yes, already proposed: #7976

> I agree, let's do this PR first, then look at further optimizations.

Awesome! So now it only needs reviews then :-)

**Comment by haukepetersen** [comment]
@everyone: can we get this PR on its way, so is anyone able to review? If we wait too long this is one of the PRs that is prone to rebasing... Thanks :-)

**Comment by miri64** [comment]
@haukepetersen: @everyone is some organization from Glasgow, if you want to mention all maintainers @RIOT-OS/maintainers is the mention you want to make ;-)

**Comment by haukepetersen** [comment]
ups :-)

**Comment by gebart** [comment]
I approve of the general idea, but I did not yet have time to do a proper review.

File: boards/common/iotlab/Makefile (hunk 1 above)
Comment (pyropeter) [reply boards/common/iotlab/Makefile]: Git does not handle renames. `git mv` just calls `git rm`, `mv` and `git add`. This is likely a glitch from gits magic rename detection code that tries to guess if files were renamed while looking at a changeset.

**Comment by PeterKietzmann** [comment]
I will try to review some parts now, combined with sporadic tests. Starting with arduino.

**Comment by PeterKietzmann** [comment]
Arduino looks okay to me and some tests with a mega2560 and duemilanove succeeded. By the way, even if it is not part of this PR I still think LED definitions and stuff should be part of the individual files.
Will check iotlab-m3 next

**Comment by PeterKietzmann** [comment]
Checked iotlab-m3,it's fine

**Comment by PeterKietzmann** [comment]
Moving msba2 didn't beak anything. Building the toolchain also still works. Would you have a look at the wiki repo and check if the old board path structure is used somewhere? 
Couple of other issues with msb* boards, but not related to your PR.

**Comment by PeterKietzmann** [comment]
Nucleo-L073 runs. I assume sooner or later we'll try to unify common things/scripts between nucleo32, nucleo64 and nucleo144. Any thoughts on this?

**Comment by PeterKietzmann** [comment]
Remote-reva success

**Comment by smlng** [comment]
it also doesn't break untouched boards such as `cc2650stk`

**Comment by haukepetersen** [comment]
> I assume sooner or later we'll try to unify common things/scripts between nucleo32, nucleo64 and nucleo144

Yes, already work in progress.

**Comment by haukepetersen** [comment]
@PeterKietzmann so from your feedback I take it that you are also happy? If so, feel free to hit the big green button :-)

**Comment by PeterKietzmann** [comment]
Unrelated issues with msb-430. Shell et al do work though.

**Comment by PeterKietzmann** [comment]
BTW you need to rebase

**Comment by haukepetersen** [comment]
oh man, seems I was to quick to merge silabs PR this morning. Will rebase now.

**Comment by haukepetersen** [comment]
rebased

**Comment by PeterKietzmann** [comment]
@cladmi would you merge this one?!

**Comment by cladmi** [comment]
I would merge it yes.

But as @kaspar030 said "We should give a change of this magnitude a little more time".
I think we can still wait at least 3 more days to have a symbolic 2 week elapsed since the PR creation to gather feedback.

It could also be an occasion to write the hierarchy in the documentation.

@gebart would that give you time to review it ? do you agree to merge it without your review or do you want a bit more time ?

**Comment by gebart** [comment]
I agree with merging even without my review as long as at least one other person has reviewed it

**Comment by PeterKietzmann** [comment]
Three days passed

**Comment by smlng** [comment]
needs rebase, my ACK holds

**Comment by haukepetersen** [comment]
and only 5 new conflicts... As stated before, waiting with this kind of PR is no fun for the author :-)

**Comment by aabadie** [comment]
> waiting with this kind of PR is no fun for the author :-)

It's no fun either for authors of PRs introducing new boards ;)

**Comment by haukepetersen** [comment]
> It's no fun either for authors of PRs introducing new boards ;)

Nope, neither the hen nor the eggs have fun :-)

**Comment by haukepetersen** [comment]
rebased

**Comment by cladmi** [comment]
Murdock is happy, no one complained.

**Comment by haukepetersen** [comment]
awesome, thanks everyone.
